### PR TITLE
HWY-197 - Testing HighwayProtocol & state::State<ClContext>

### PR DESCRIPTION
**Pull request by xcthulhu**
JIRA: https://casperlabs.atlassian.net/browse/HWY-197

Description:
   - A basic test for `highway_core::state::State<ClContext>` is introduced, in concert with a `HighwayProtocol` instance
   - Some common test infrastructure has been factored out

### CODE REVIEW

File: node/src/components/consensus/era_supervisor/tests.rs [hunk 1]
```diff
@@ -47,6 +56,7 @@ enum Event {
     Storage(StorageRequest),
     #[from]
     ContractRuntime(ContractRuntimeRequest),
+    // ConsensusRequest(ConsensusRequest),
```
Comment (xcthulhu) [inline node/src/components/consensus/era_supervisor/tests.rs:59]: @afck Was this intentional?

**Comment by xcthulhu** [comment]
As discussed with @afck , I'm going to break this up into two PRs

File: node/src/components/consensus/era_supervisor/tests.rs (hunk 1 above)
Comment (afck) [reply node/src/components/consensus/era_supervisor/tests.rs:59]: I guess this can be removed entirely, actually.

**Comment by xcthulhu** [comment]
Blocked by #560 (since this PR builds on that branch)

**Comment by goral09** [comment]
I don't think this ticket is visible on your board.

File: node/src/components/consensus/protocols/highway/tests.rs [hunk 2]
```diff
@@ -81,3 +89,52 @@ fn test() {
...
+        creator,
+        instance_id: ClContext::hash(INSTANCE_ID_DATA),
+        value: None,
+        seq_number,
+        timestamp: 0.into(),
+        round_exp: 0,
+        endorsed: BTreeSet::new(),
+    };
+    let alice_keypair: Keypair =
+        Rc::new(SecretKey::ed25519_from_bytes(&*ALICE_RAW_SECRET).unwrap()).into();
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs:113]: `ALICE_SECRET_KEY` doesn't implement `Copy`, so I need to use the underlying secret bytes to construct the appropriate `KeyPair` here.  But maybe a better way would be to construct `ALICE_KEYPAIR` as a `lazy_static!` and then provide `ALICE_{SECRET,PUBLIC}_KEY` based on the `KeyPair`...

File: node/src/components/consensus/tests/utils.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    utils::Loadable,
+    Chainspec,
+};
+
+lazy_static! {
+    pub static ref ALICE_RAW_SECRET: Vec<u8> = vec![0; SecretKey::ED25519_LENGTH];
+    pub static ref ALICE_SECRET_KEY: SecretKey =
+        SecretKey::ed25519_from_bytes(&*ALICE_RAW_SECRET).unwrap();
+    pub static ref ALICE_PUBLIC_KEY: PublicKey = PublicKey::from(ALICE_SECRET_KEY.deref());
+    pub static ref BOB_RAW_SECRET: Vec<u8> = vec![1; SecretKey::ED25519_LENGTH];
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/tests/utils.rs:17]: I kept this one for symmetry with the `ALICE_RAW_SECRET`.  I'll look into using a `Keypair` however.

File: node/src/components/consensus/protocols/highway/tests.rs [hunk 4]
```diff
@@ -1,67 +1,75 @@
...
+        traits::Context,
+        HighwayProtocol,
     },
-    crypto::asymmetric_key::{PublicKey, SecretKey},
+    crypto::asymmetric_key::SecretKey,
     testing::TestRng,
-    utils::Loadable,
+    types::Timestamp,
 };
+use std::collections::BTreeSet;
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs:28]: :facepalm:

**Comment by xcthulhu** [comment]
@afck thank you very much for the review!  I don't have time this week to take care of this but I'll tackle it as soon as I get back.

File: node/src/components/consensus/protocols/highway/tests.rs [hunk 5]
```diff
@@ -81,3 +89,52 @@ fn test() {
...
+    match &opt_protocol_outcome {
+        None => panic!("We just checked that effects has length 1!"),
+        Some(ProtocolOutcome::InvalidIncomingMessage(invalid_msg, offending_sender, err)) => {
+            assert_eq!(
+                invalid_msg, &msg,
+                "Invalid message is not message that was sent."
+            );
+            assert_eq!(offending_sender, &sender, "Unexpected sender.");
+            assert!(format!("{:?}", err).starts_with(
+                "The vertex contains an invalid unit: `The round length exponent is less than the minimum allowed by the chain-spec.`"),
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs:133]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/385aaf0fe84ec1f678200497ef84cf491ae3c834

File: node/src/components/consensus/protocols/highway/tests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/385aaf0fe84ec1f678200497ef84cf491ae3c834

File: node/src/components/consensus/protocols/highway/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/4eaa37f731ded43efe02b6462ff785c955482348

File: node/src/components/consensus/tests/utils.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        .into_iter()
+        .map(|(pk, stake)| {
+            let motes = Motes::new(stake.into());
+            GenesisAccount::new(pk.into(), pk.to_account_hash(), motes, motes)
+        })
+        .collect();
+    chainspec.genesis.timestamp = Timestamp::now();
+    chainspec.genesis.highway_config.genesis_era_start_timestamp = chainspec.genesis.timestamp;
+
+    // Every era has exactly three blocks.
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/tests/utils.rs:37]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/b4dcd427f7c762a892657fcce31003098dcf58fd

**Comment by xcthulhu** [comment]
As @afck has pointed out to me, I need to replace my use of `lazy_static!` in this code...

File: node/src/components/consensus/protocols/highway/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/4f56c814ec0d4b769d07c3bb8082a149b36bd9de

File: node/src/components/consensus/tests/utils.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/components/consensus/tests/utils.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/4f56c814ec0d4b769d07c3bb8082a149b36bd9de

File: node/src/components/consensus/tests/utils.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    chainspec.genesis.highway_config.minimum_era_height = 2;
```
Comment (afck) [inline node/src/components/consensus/tests/utils.rs:34]: Wait, those are _two_ blocks, not three! :smile: 
So the comment above is wrong.

Comment (afck) [reply node/src/components/consensus/tests/utils.rs:34]: (The off-by-one error is fixed on master, I think.)

Comment (xcthulhu) [reply node/src/components/consensus/tests/utils.rs:34]: Fixed in https://github.com/CasperLabs/casper-node/pull/559/commits/d5a02299c8fa778c3458eb6fa38730812b48fe13

**Review by afck — APPROVED** [review]

**Comment by xcthulhu** [comment]
bors merge

**Pull request merged**